sycl/source/detail/persistent_device_code_cache.cpp

```diff
@@ -0,0 +1,408 @@
+//==---------- persistent_device_code_cache.cpp -----------------*- C++-*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <detail/config.hpp>
+#include <detail/device_impl.hpp>
+#include <detail/persistent_device_code_cache.hpp>
+#include <detail/plugin.hpp>
+
+#if defined(__SYCL_RT_OS_LINUX)
+#include <unistd.h>
+#else
+#include <direct.h>
+#include <io.h>
+#endif
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+/* This is temporary solution until std::filesystem is available when SYCL RT
+ * is moved to c++17 standard*/
+
+/* Create directory recursively */
+int makeDir(const char *Dir) {
+  assert((Dir != nullptr) && "Passed null-pointer as directory name.");
+  if (isPathPresent(Dir))
+    return 0;
+
+  std::string Path{Dir}, CurPath;
+  size_t pos = 0;
+
+  do {
+    pos = Path.find_first_of("/\\", ++pos);
+    CurPath = Path.substr(0, pos);
+#if defined(__SYCL_RT_OS_LINUX)
+    auto Res = mkdir(CurPath.c_str(), 0777);
+#else
+    auto Res = _mkdir(CurPath.c_str());
+#endif
+    if (Res && errno != EEXIST)
+      return Res;
+  } while (pos != std::string::npos);
+  return 0;
+}
+
+const char LockCacheItem::LockSuffix[] = ".lock";
+LockCacheItem::LockCacheItem(const std::string &Path)
+    : FileName(Path + LockSuffix) {
+  int fd;
+  auto Start = std::chrono::high_resolution_clock::now();
+
+  while ((fd = open(FileName.c_str(), O_CREAT | O_EXCL, S_IWRITE)) == -1) {
+    // if lock file is not created unblock the thread
+    if (std::chrono::high_resolution_clock::now() - Start >
+        std::chrono::microseconds(100)) {
+      return;
+    }
+    std::this_thread::yield();
+  }
+  close(fd);
+  Owned = true;
+}
+
+LockCacheItem::~LockCacheItem() {
+  if (Owned) {
+    auto Start = std::chrono::high_resolution_clock::now();
+    while (std::remove(FileName.c_str())) {
+      // if lock file is not cleaned unblock the thread
+      if (std::chrono::high_resolution_clock::now() - Start >
+          std::chrono::microseconds(10))
```

Comment:
What is the scenario where failure to remove a file can succeed after retry? I'd suggest to remove the loop.
Otherwise (if there is a scenario) - please make `10` a named constant. BTW, I think `10` microseconds might not be long enough for any file I/O call - did you test it?

Reply:
done